### ISSUE TITLE
("No module named 'pyexiv2'") How to install pyexiv2? 

### ISSUE THREAD

**Issue by pxssy** [Nandaka/PixivUtil2#1004]
I see you require >Exiv2 (https://exiv2.org/) - used for writing XMP metadata (if enabled).

but i can't find exactly how to install this thing. 

Can someone give me a few directions on how/where to exactly put this thing? I'm running the source code with python 3.9.6

Is https://github.com/LeoHsiao1/pyexiv2 the same as the Exiv2? This one has a pip install, which i think i can try but i'm not exactly sure.

**Comment by Nandaka** [comment]
need to ask @Baa14453 from PR #980

**Comment by Baa14453** [comment]
> I see you require >Exiv2 (https://exiv2.org/) - used for writing XMP metadata (if enabled).
> 
> but i can't find exactly how to install this thing.
> 
> Can someone give me a few directions on how/where to exactly put this thing? I'm running the source code with python 3.9.6
> 
> Is https://github.com/LeoHsiao1/pyexiv2 the same as the Exiv2? This one has a pip install, which i think i can try but i'm not exactly sure.

Hi,
You only need the Pyexiv2 library if you want to use the options `writeImageXMP` or `writeImageXMPPerImage`.
You can install it from pip, there's a commented out line in the requirements.txt

Looking at the installation for Pyexiv2 you don't need Exiv2 to be installed separately, maybe I thought you did at the time not sure why I added it to the Readme if it wasn't required...

Anyway, there's an FAQ in the installation instructions for Pyexiv2 that has some helpful pointers and common errors: https://github.com/LeoHsiao1/pyexiv2/blob/master/docs/Tutorial.md#installation

**Comment by pxssy** [comment]
I understand there's a pip install for Pyexiv2. I'm just wondering if i can use Pyexiv2, or i must use the Exiv2 as stated in the requirements/dependent software.

**Comment by Baa14453** [comment]
You just need Pyexiv2
Pyexiv2 is a wrapper for Exiv2, I thought you needed both of them but pretty sure I was wrong.

**Comment by pxssy** [comment]
Still doesn't seem to work

`Traceback (most recent call last):
  File "C:\PixivUtil2-20210822\PixivImageHandler.py", line 292, in process_image
    image.WriteXMP(info_filename + ".xmp")
  File "C:\PixivUtil2-20210822\PixivImage.py", line 457, in WriteXMP
    info = pyexiv2.ImageMetadata(filename)
AttributeError: module 'pyexiv2' has no attribute 'ImageMetadata'
Error at process_image(): 92079661
Exception: (<class 'AttributeError'>, AttributeError("module 'pyexiv2' has no attribute 'ImageMetadata'"), <traceback object at 0x000002137AE88100>)
Dumping html to: Error medium page for image 92079661.html
Stuff happened, trying again after 2 second (1)
Traceback (most recent call last):
  File "C:\PixivUtil2-20210822\PixivArtistHandler.py", line 161, in process_member
    result = PixivImageHandler.process_image(caller,
  File "C:\PixivUtil2-20210822\PixivImageHandler.py", line 292, in process_image
    image.WriteXMP(info_filename + ".xmp")
  File "C:\PixivUtil2-20210822\PixivImage.py", line 457, in WriteXMP
    info = pyexiv2.ImageMetadata(filename)
AttributeError: module 'pyexiv2' has no attribute 'ImageMetadata'
Error at process_member(): (<class 'AttributeError'>, AttributeError("module 'pyexiv2' has no attribute 'ImageMetadata'"), <traceback object at 0x000002137AFAB800>) Member Id: 32045169`

**Comment by registered-to-post-this** [comment]
TL;DR: It only works on Linux.

The module you need is actually called py3exiv2. The version in the pip repo (https://pypi.org/project/py3exiv2/) doesn't work on Windows 10 x64, see error log: 

<details>
  <summary>Click to expand error log</summary>
  
  ```PS C:\Users\admin\Desktop> python -m pip install py3exiv2
Collecting py3exiv2
  Using cached py3exiv2-0.9.3.tar.gz (36 kB)
Building wheels for collected packages: py3exiv2
  Building wheel for py3exiv2 (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: 'C:\Program Files\Python310\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\admin\\AppData\\Local\\Temp\\pip-install-phptlggs\\py3exiv2_0ce89e177735453daa26247589a9470f\\setup.py'"'"'; __file__='"'"'C:\\Users\\admin\\AppData\\Local\\Temp\\pip-install-phptlggs\\py3exiv2_0ce89e177735453daa26247589a9470f\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\admin\AppData\Local\Temp\pip-wheel-udlolrgf'
       cwd: C:\Users\admin\AppData\Local\Temp\pip-install-phptlggs\py3exiv2_0ce89e177735453daa26247589a9470f\
  Complete output (23 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.10
  creating build\lib.win-amd64-3.10\pyexiv2
  copying src\pyexiv2\exif.py -> build\lib.win-amd64-3.10\pyexiv2
  copying src\pyexiv2\iptc.py -> build\lib.win-amd64-3.10\pyexiv2
  copying src\pyexiv2\metadata.py -> build\lib.win-amd64-3.10\pyexiv2
  copying src\pyexiv2\preview.py -> build\lib.win-amd64-3.10\pyexiv2
  copying src\pyexiv2\utils.py -> build\lib.win-amd64-3.10\pyexiv2
  copying src\pyexiv2\xmp.py -> build\lib.win-amd64-3.10\pyexiv2
  copying src\pyexiv2\__init__.py -> build\lib.win-amd64-3.10\pyexiv2
  running build_ext
  building 'libexiv2python' extension
  creating build\temp.win-amd64-3.10
  creating build\temp.win-amd64-3.10\Release
  creating build\temp.win-amd64-3.10\Release\src
  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -IC:\Program Files\Python310\include -IC:\Program Files\Python310\Include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\include -IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt /EHsc /Tpsrc/exiv2wrapper.cpp /Fobuild\temp.win-amd64-3.10\Release\src/exiv2wrapper.obj -g
  cl : Command line warning D9002 : ignoring unknown option '-g'
  exiv2wrapper.cpp
  C:\Users\admin\AppData\Local\Temp\pip-install-phptlggs\py3exiv2_0ce89e177735453daa26247589a9470f\src\exiv2wrapper.hpp(32): fatal error C1083: Cannot open include file: 'exiv2/exiv2.hpp': No such file or directory
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
  ----------------------------------------
  ERROR: Failed building wheel for py3exiv2
  Running setup.py clean for py3exiv2
Failed to build py3exiv2
Installing collected packages: py3exiv2
    Running setup.py install for py3exiv2 ... error
    ERROR: Command errored out with exit status 1:
     command: 'C:\Program Files\Python310\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\admin\\AppData\\Local\\Temp\\pip-install-phptlggs\\py3exiv2_0ce89e177735453daa26247589a9470f\\setup.py'"'"'; __file__='"'"'C:\\Users\\admin\\AppData\\Local\\Temp\\pip-install-phptlggs\\py3exiv2_0ce89e177735453daa26247589a9470f\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\admin\AppData\Local\Temp\pip-record-5qkapkkt\install-record.txt' --single-version-externally-managed --compile --install-headers 'C:\Program Files\Python310\Include\py3exiv2'
         cwd: C:\Users\admin\AppData\Local\Temp\pip-install-phptlggs\py3exiv2_0ce89e177735453daa26247589a9470f\
    Complete output (23 lines):
    running install
    running build
    running build_py
    creating build
    creating build\lib.win-amd64-3.10
    creating build\lib.win-amd64-3.10\pyexiv2
    copying src\pyexiv2\exif.py -> build\lib.win-amd64-3.10\pyexiv2
    copying src\pyexiv2\iptc.py -> build\lib.win-amd64-3.10\pyexiv2
    copying src\pyexiv2\metadata.py -> build\lib.win-amd64-3.10\pyexiv2
    copying src\pyexiv2\preview.py -> build\lib.win-amd64-3.10\pyexiv2
    copying src\pyexiv2\utils.py -> build\lib.win-amd64-3.10\pyexiv2
    copying src\pyexiv2\xmp.py -> build\lib.win-amd64-3.10\pyexiv2
    copying src\pyexiv2\__init__.py -> build\lib.win-amd64-3.10\pyexiv2
    running build_ext
    building 'libexiv2python' extension
    creating build\temp.win-amd64-3.10
    creating build\temp.win-amd64-3.10\Release
    creating build\temp.win-amd64-3.10\Release\src
    C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -IC:\Program Files\Python310\include -IC:\Program Files\Python310\Include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\include -IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt /EHsc /Tpsrc/exiv2wrapper.cpp /Fobuild\temp.win-amd64-3.10\Release\src/exiv2wrapper.obj -g
    cl : Command line warning D9002 : ignoring unknown option '-g'
    exiv2wrapper.cpp
    C:\Users\admin\AppData\Local\Temp\pip-install-phptlggs\py3exiv2_0ce89e177735453daa26247589a9470f\src\exiv2wrapper.hpp(32): fatal error C1083: Cannot open include file: 'exiv2/exiv2.hpp': No such file or directory
    error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
    ----------------------------------------
ERROR: Command errored out with exit status 1: 'C:\Program Files\Python310\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\admin\\AppData\\Local\\Temp\\pip-install-phptlggs\\py3exiv2_0ce89e177735453daa26247589a9470f\\setup.py'"'"'; __file__='"'"'C:\\Users\\admin\\AppData\\Local\\Temp\\pip-install-phptlggs\\py3exiv2_0ce89e177735453daa26247589a9470f\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\admin\AppData\Local\Temp\pip-record-5qkapkkt\install-record.txt' --single-version-externally-managed --compile --install-headers 'C:\Program Files\Python310\Include\py3exiv2' Check the logs for full command output.
  ```
</details>

There's a fork with Windows support but it doesn't work either, here's as far as I got with it:

* Download Python 3.9.*  (any other version will fail to build) and follow https://github.com/auphofBSF/py3exiv2
* You have to reboot before Python will recognize any changes to PATH entries or environment variables
* If you run "pip install -r requirements.txt" in the PixivUtil2 folder after installing py3exiv2 manually, leave #py3exiv2 commented out in requirements.txt or pip will uninstall it and attempt to replace it with the pip repo version

But this doesn't work either, here's what will happen when you try and run it

<details>
  <summary>Click to expand error log</summary>
  
  ```Traceback (most recent call last):
  File "C:\Users\admin\Desktop\PixivUtil2-master\PixivImageHandler.py", line 292, in process_image
    image.WriteXMP(info_filename + ".xmp")
  File "C:\Users\admin\Desktop\PixivUtil2-master\PixivImage.py", line 441, in WriteXMP
    import pyexiv2
  File "c:\users\admin\desktop\py3exiv2-master\src\pyexiv2\__init__.py", line 60, in <module>
    import libexiv2python
ImportError: DLL load failed while importing libexiv2python: The specified module could not be found.
Error at process_image(): 92079661
Exception: (<class 'ImportError'>, ImportError('DLL load failed while importing libexiv2python: The specified module could not be found.'), <traceback object at 0x0000014ADBDC5A80>)
Dumping html to: Error medium page for image 92079661.html
Unknown Error, please check the log file: (<class 'ImportError'>, ImportError('DLL load failed while importing libexiv2python: The specified module could not be found.'), <traceback object at 0x0000014ADBB4E800>)
  ```
</details>

But I don't have any DLL files. After running "pip install py3exiv2-master", there's a "C:\Program Files\Python39\Lib\site-packages\libexiv2python.cp39-win_amd64.pyd" - but unfortunately renaming it to "libexiv2python.pyd" or "libexiv2python.dll" doesn't work. Which is too bad, that would have been funny if it did.


I don't know anything about Python - this is all just googling and trial and error - so maybe this is actually an easy fix. I have no idea.

**Comment by Nandaka** [comment]
try this one: https://github.com/LeoHsiao1/pyexiv2/releases/tag/v2.7.0 

Looks like it got 64bit support on windows.

**Comment by Nandaka** [comment]
rewrite the pyexiv2 part to support pyexiv2-2.7.0-cp39-none-win_amd64.whl (only have 64bit version here)
